library/src/main/java/com/yausername/youtubedl_android/YoutubeDL.kt

```diff
@@ -16,84 +46,184 @@ import kotlin.collections.set
+    @Throws(YoutubeDLException::class, MissingPlugin::class)
+    /**
+     * Initializes the library. This method should be called before any other method.
+     * @param appContext the application context
+     */
+    fun init(appContext: Context, callback: pluginDownloadCallback? = null) {
         if (initialized) return
-        val baseDir = File(appContext.noBackupFilesDir, baseName)
-        if (!baseDir.exists()) baseDir.mkdir()
-        val packagesDir = File(baseDir, packagesRoot)
-        binDir = File(appContext.applicationInfo.nativeLibraryDir)
-        pythonPath = File(binDir, pythonBinName)
-        ffmpegPath = File(binDir, ffmpegBinName)
-        val pythonDir = File(packagesDir, pythonDirName)
-        val ffmpegDir = File(packagesDir, ffmpegDirName)
-        val aria2cDir = File(packagesDir, aria2cDirName)
-        val ytdlpDir = File(baseDir, ytdlpDirName)
-        ytdlpPath = File(ytdlpDir, ytdlpBin)
+
+        val baseDir = File(appContext.noBackupFilesDir, LIBRARY_NAME)
+
+        assertAndCreate(baseDir)
+
+        val packagesDir = File(baseDir, PACKAGES_ROOT_NAME)
+
+        binariesDirectory = File(appContext.applicationInfo.nativeLibraryDir) //Here are all the binaries provided by the jniLibs folder
+
+        pythonPath = File(binariesDirectory, PYTHON_BINARY_NAME)
+        ffmpegPath = File(binariesDirectory, FFMPEG_BINARY_NAME)
+
+        // Create plugins packages directory (where they are extracted)
+        val pythonDir = File(packagesDir, PYTHON_DIRECTORY_NAME)
+        val ffmpegDir = File(packagesDir, FFMPEG_DIRECTORY_NAME)
+        val aria2cDir = File(packagesDir, ARIA2C_DIRECTORY_NAME)
+
+        val ytdlpDir = File(baseDir, YTDLP_DIRECTORY_NAME)
+        ytdlpPath = File(ytdlpDir, YTDLP_BINARY_NAME)
+
+        val installedPlugins = checkInstalledPlugins(appContext)
+
+        // Set environment variables
         ENV_LD_LIBRARY_PATH = pythonDir.absolutePath + "/usr/lib" + ":" +
                 ffmpegDir.absolutePath + "/usr/lib" + ":" +
                 aria2cDir.absolutePath + "/usr/lib"
         ENV_SSL_CERT_FILE = pythonDir.absolutePath + "/usr/etc/tls/cert.pem"
         ENV_PYTHONHOME = pythonDir.absolutePath + "/usr"
+
+        assertPlugins(appContext, installedPlugins) { plugin, progress ->
+            callback?.invoke(plugin, progress)
+        }
+
         initPython(appContext, pythonDir)
-        init_ytdlp(appContext, ytdlpDir)
+        initYtdlp(appContext, ytdlpDir)
+
         initialized = true
     }
 
+    /**
+     * Asserts that the plugins are installed and downloads them if they are missing
+     * @param appContext the application context
+     * @param downloadedPlugins the downloaded plugins
+     * @param callback a callback that will be called with the plugin and the progress of the download
+     */
+    @Throws(MissingPlugin::class)
+    fun assertPlugins(appContext: Context, downloadedPlugins: DownloadedPlugins, callback: pluginDownloadCallback?) {
```

Comment:
A function called `assert` should do nothing other than assertions i think? maybe make it returns a boolean or throws an exception, then leave the rest to the caller

Reply:
I've renamed the function. You tell me if it's okay